### PR TITLE
WT-9040 Add configuration to allow a transaction to skip eviction help.

### PR DESCRIPTION
**Pull request by sueloverso**

### CODE REVIEW

**Comment by sueloverso** [comment]
@keitharnoldsmith and @ddanderson  I am marking this as a draft for Storage Execution to try out but I'd be interested in any review comments for the code. If this is an acceptable API, it is not really a draft. But we'll leave it as such until we think we want it.

File: dist/api_data.py
Comment (ddanderson) [inline dist/api_data.py:1609]: I assume we don't want "before or", or is there time travel involved? :-)

File: src/txn/txn.c
Comment (ddanderson) [inline src/txn/txn.c:731]: typo

**Review by ddanderson — APPROVED** [review]
LGTM modulo a couple of small items in comments.

File: dist/api_data.py
Comment (sueloverso) [reply dist/api_data.py:1609]: It is accurate. In `begin_transaction` just before we set `WT_TXN_RUNNING` it calls `wt_cache_eviction_check`. So threads can be pulled into helping with eviction before the transaction starts. Similarly on the commit and rollback, there is a call just after `wt_txn_release`.

The "during" part is the less common path because if one begins a transaction and then lets it sit idle we also call `wt_txn_idle_cache_check`. So the setting on `begin_transaction` will also disable that call during the transaction.

File: src/txn/txn.c
Comment (sueloverso) [reply src/txn/txn.c:731]: Good catch, I have no idea how that got there.

File: dist/api_data.py
Comment (keitharnoldsmith) [reply dist/api_data.py:1609]: At the risk of splitting semantic hairs...

I would think that a typical reader/programmer just thinks of the transaction as starting as soon as their program calls `begin_transaction()`, and isn't aware of the finer distinction of where we check for eviction compared to where we set `WT_TXN_RUNNING`. So from that perspective it is all "during" the transaction.

File: dist/api_data.py
Comment (keitharnoldsmith) [inline dist/api_data.py:1608]: Do we want these options to be `undoc=True`? This is option has sharp edges, so we might not want the typical user (if there is such a beast) using it.

File: dist/stat_data.py
Comment (keitharnoldsmith) [inline dist/stat_data.py:1007]: Maybe `number of times application eviction was skipped`? 

I think it will be confusing to just call this `eviction` as people could confuse it with the other stats that use the term "eviction" more generally. E.g., "leaf pages split during eviction". So `application eviction` seems better. 

`for the transaction` doesn't really add anything for me and is a bit confusing. Which transaction is *the* transaction? I suppose maybe it could say `for a transaction` or `during transaction API call`

**Review by keitharnoldsmith — COMMENTED** [review]
I've added a few more comments about the documentation.

Since this is intended as a draft, I'm going to hold off on the LGTM to make sure we (in the general sense; doesn't have to include me) discuss the API to make absolutely certain this is what we want.

File: dist/api_data.py
Comment (sueloverso) [reply dist/api_data.py:1609]: Good point. I'll get rid of "before"

File: dist/api_data.py
Comment (sueloverso) [reply dist/api_data.py:1608]: Yes, I will make them all undoc.

File: dist/stat_data.py
Comment (sueloverso) [reply dist/stat_data.py:1007]: I added 'application'. I'm also not happy with 'skipped' as it implies the code made a choice to skip it based on some conditions. Maybe 'turned off'?

**Comment by sueloverso** [comment]
> I think for at least the begin_transaction, and probably for the commit_transaction version, the right solution is to return a WT_ROLLBACK (the option should therefore be disallowed on prepared transactions, since they aren't allowed to return WT_ROLLBACK).

@agorrod  While the code change isn't hard inside `wt_cache_eviction_check` that solution is not going to work and will make the code harder to understand. Both `commit` and `rollback` call the eviction check after resolving the transaction so both calls ignore any error from `wt_cache_eviction_check`. So WT_ROLLBACK will be ignored.

In `begin` we also ignore the error because it also cannot return WT_ROLLBACK. There is this comment:
```
        /*
         * Stall here if the cache is completely full. Eviction check can return rollback, but the
         * WT_SESSION.begin_transaction API can't, continue on.
         */
        WT_RET_ERROR_OK(__wt_cache_eviction_check(session, false, true, NULL), WT_ROLLBACK);
```
so returning WT_ROLLBACK for this situation in any API because it is not allowed.

**Pull request closed without merging**
